### PR TITLE
#347 Only pass fields that are known to be in the User Model

### PR DESCRIPTION
**Pull request by BillHeaton**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #345, #346, #347

Adding any field to the registration form fails because the container doesn't know what to do with the extra field(s). The solution is to only pass the fields that the user model expects to show up.

### CODE REVIEW

**Comment by tonydspaniard** [comment]
hi @BillHeaton whats the issue this PR solves? 

According to the documentation you can override any model class (i.e. User), when the user is overridden on the container, when you call `->make(User::class)` is in fact, returning your class (the one you override, thus containing the new attributes)

**Comment by tonydspaniard** [comment]
I understand the PR but I think it can be solved in two ways: 

- one, using https://www.php.net/manual/en/function.array-intersect-key.php 
```
$user = $this->make(User::class, []); 
$user->attributes = array_intersect_key($form->attributes, $user->attributes); // not tested 
$user->setScenario('register');
```
- two, overriding the `RegistrationController` class and fix your issue as it relates to the form you modified (adding captcha for example) 

I don't think that hardcoding the assignment would be good. In order to allow this PR I would use the option one, but you would have to update your code and also, please, add the credits to the CHANGELOG.md 

I highly appreciate the time you spent to work on this.

**Comment by BillHeaton** [comment]
@tonydspaniard  Unfortunately that intersect doesn't work because the fields haven't been filled in.  However, this kludge appears to work.  Perhaps we have room for improvement?

```
$user = $this->make(User::class, [] );  // Get User
$fields = array_intersect_key($form->attributes, $user->attributes);
$fields['password'] = $form['password'];         //  Gets changed to password_hash
$user = $this->make(User::class, [], $fields );  // Do the class for real
```

**Comment by BillHeaton** [comment]
@tonydspaniard Have updated the guide, recoveryController, and credits to reflect your comments.

**Comment by tonydspaniard** [comment]
Thanks! I'll review and merge tomorrow

File: docs/helpful-guides/how-to-use-recaptcha-widget.md [hunk 1]
```diff
@@ -30,34 +30,47 @@ Once you have the API site key you will also be displayed a secret key. You have
...
 ```
 
+
 Overriding the View
 -------------------
 
 Create a new file and name it `request.php` and add it in `@app/views/user/recovery`. Add the captcha widget to it: 
 
-```php 
+```
```
Comment (maxxer) [inline docs/helpful-guides/how-to-use-recaptcha-widget.md]: Why did you remove the lang tag?

Comment (BillHeaton) [reply docs/helpful-guides/how-to-use-recaptcha-widget.md]: It's on the next line as ````<?php````

Comment (maxxer) [reply docs/helpful-guides/how-to-use-recaptcha-widget.md]: Does it work anyway? Because they're different things: the removed one is a Markdown syntax, while the one below is programming

Comment (BillHeaton) [reply docs/helpful-guides/how-to-use-recaptcha-widget.md]: Github markdown extension I didn't know about.  Works either way.  It appears that the syntax coloring still happens.

**Comment by srakl** [comment]
what is the solution to get it to work on user registration with custom fields? (firstname, lastname) ?
